### DISCUSSION TITLE
Retransmissions in iperf3 vs retransmissions in wireshark

### DISCUSSION THREAD

**Discussion by johnerino123** [esnet/iperf#1627] [Q&A]
Why would tcpdump output (listening on client interface) show much less retransmissions than what iperf shows?

**Comment by davidBar-On** [comment]
iperf3 tcp retransmissions value is taken from the [tcpi_total_retrans](https://docs.huihoo.com/doxygen/linux/kernel/3.7/structtcp__info.html#a14efa163e515a0b79b03c12dae1e02df) of [tcp_info](https://docs.huihoo.com/doxygen/linux/kernel/3.7/include_2uapi_2linux_2tcp_8h_source.html).  `tcp_info` includes 3 TCP retransmission related fields: `tcpi_retransmits`, `tcpi_retrans`, `tcpi_total_retrans`.  [This comment](https://narkive.com/1dzGbgsZ.3) explains that `tcpi_retransmits` counts only timeout based retransmissions, while `tcpi_retrans` counts also fast-retransmits, etc.  I assume that `tcpi_total_retrans` counts the same retransmissions as `tcpi_retrans` (I didn't find where this is documented).

If tcpdump counts only the timeout based retransmissions, then it is expected to be less than `tcpi_total_retrans` which is reported by iper3.

**Reply by ddimopou** [reply]
Hello, 

I face the same behavior as the one reported here (less retransmissions captured on the client interface than the number reported by iperf3). From [https://github.com/esnet/iperf/issues/343](url), it seems that if the total number of retransmissions reported by iperf includes also the retransmission attempts, then the reason why iperf shows much more retransmissions is that something goes wrong before the packets are placed on the interface. 

What could be a possible cause of having failures before packets reach the interface? (could it be CPU limits? / high IO?)

In my case I am testing a MPTCP scenario on Mininet, but I'm just curious to understand the reason behind this discrepancy.
The client sends traffic over two interfaces (10.0.0.2 & 10.0.1.2), and the server receives packets via a single interface (10.0.2.2).

I'd like to capture retransmissions in my tests, and I'm not sure which value is the valid one (i.e. the one coming from iperf or those measured on the wire). It seems that tcpdump will measure retransmission packets that made it to reach the interface and were actually re-transmitted.

Attaching the dump files captured on the client (cl.pcap) and the server side (serv.pcap), as well as the iperf3 client output in json format.

Filtering with "tcp.analysis.retransmission" in wireshark seems to count all types of retransmissions (also spurious and fast). The server side dump captures 49, the client 37 (also not sure why the client reports less than the server -- looks like the client couldn't capture its own retransmitted packets while these were actually sent and captured by the server end), and iperf reports 136. 

However, it seems that the majority of retransmissions reported through iperf occur at the beginning of the test. Could this burst during i.e. tcp slow start phase justify some retransmission attempts' failures?  


Please correct any of my assumptions above.
Your help to better understand this would be much appreciated. 

Thanks,
Dimitris

![cl](https://github.com/user-attachments/assets/eda6aded-71b6-4ac2-8db2-32452287a510)

![serv](https://github.com/user-attachments/assets/e045c143-1b50-4628-9c23-ce85a3ee00e0)

Iperf3 output on client side:  [cl.txt](https://github.com/user-attachments/files/18066336/cl.txt)
